Added humble branch

routiful opened the pull request:
This framework is pretty excited! I have tested this on ROS 2 humble ant it works well!
Since ROS 2 API was changed, however, it is needed to modify some files.
I would expect someone want to run this on humble. I added comments for what I changed.

1. https://github.com/42dot/foros/commit/2f877f5f51f3b8523859add2c0b530e9d46720b1
[rclcpp/client.hpp/SharedFuture()](https://github.com/ros2/rclcpp/blob/7f575103d8fe06285bd46474a0324c7dfbb2a2f0/rclcpp/include/rclcpp/client.hpp#L408-L415) was deprecated. 
So I added a conversion.

2. https://github.com/42dot/foros/commit/5a427e793677841f9780c245ece675c7fac2a37a
[rclcpp/node_interfaces/node_base/get_callback_group()](https://github.com/ros2/rclcpp/blob/7f575103d8fe06285bd46474a0324c7dfbb2a2f0/rclcpp/include/rclcpp/node_interfaces/node_base.hpp#L89-L110) wasn't supported on humble. 
I added deprecated log on that function and deleted a declared function in cpp.

3. https://github.com/42dot/foros/commit/2c8c77df66c510eeb821b6dc165c2f5e4f55484b
[rclcpp/publisher/publisher()](https://github.com/ros2/rclcpp/blob/7f575103d8fe06285bd46474a0324c7dfbb2a2f0/rclcpp/include/rclcpp/publisher.hpp#L235-L273) was changed. This function hasn't be permitted overriding. 
I deleted an override option in cluster_node_publisher.cpp.

Thank you for sharing this project :) and hope to see you in ROSCon 2022.

Commit:
foros: deleted override options of publisher functions

Signed-off-by: Taehun Lim <thlim@robotis.com>

Commit:
foros: added log to deprecated ros2 api

Signed-off-by: Taehun Lim <thlim@robotis.com>

Commit:
foros: modified return type

Signed-off-by: Taehun Lim <thlim@robotis.com>

huchijwk commented on foros/include/akit/failover/foros/cluster_node.hpp:
```diff
@@ -138,7 +138,9 @@ class ClusterNode : public std::enable_shared_from_this<ClusterNode>,
    * \return List of callback groups in the node.
+   * \deprecated
    */
+  [[deprecated("get_callback_groups(): Not included in humble")]]
```
`deprecated` keyword means "the use is allowed, but discouraged"
Since you removed the definition of this function, users can faced on build break when they use it.

Anyhow, this interface has been replaced with for_each_callback_group.

See https://github.com/ros2/rclcpp/pull/1723.
You can import implementations of lifecycle_node.hpp and lifecycle_node.cpp into cluster_node.hpp and cluster_node.cpp.

routiful replied:
Thank you for kind comments :) I deleted `get_callback_group()` and added `for_each_callback_group()`.

huchijwk commented on foros/src/cluster_node.cpp:
```diff
@@ -119,11 +119,6 @@ rclcpp::CallbackGroup::SharedPtr ClusterNode::create_callback_group(
       group_type, automatically_add_to_executor_with_node);
 }
 
-const std::vector<rclcpp::CallbackGroup::WeakPtr>
```
This interface has been replaced with `for_each_callback_group`.

See https://github.com/ros2/rclcpp/pull/1723.
You can import implementations of lifecycle_node.hpp and lifecycle_node.cpp into cluster_node.hpp and cluster_node.cpp.

routiful replied:
I imported an implementation about `for_each_callback_group()` to cluster_node.cpp.

huchijwk reviewed (changes requested):
Many thanks for your contribution. :)

1. From the Humble, `get_callback_group()` is replaced with `for_each_callback_group()`. Please refer to inline comments and https://github.com/ros2/rclcpp/pull/1723.
2. Please use a commit message format similar to the previous commits.
    - module or category: blan blah (all lower case)
    - ex) foros: delete override keyward from publisher()

see you at ROSCon

Commit:
foros: deleted a method

Signed-off-by: Taehun Lim <thlim@robotis.com>

Commit:
foros: added new method

Signed-off-by: Taehun Lim <thlim@robotis.com>

routiful commented:
> 1. From the Humble, `get_callback_group()` is replaced with `for_each_callback_group()`. Please refer to inline comments and [ros2/rclcpp#1723](https://github.com/ros2/rclcpp/pull/1723).

Appreciated your comments. I deleted `get_callback_group()` and added `for_each_callback_group()` with inline comments to a hpp file. Also, I imported an implementation about `for_each_callback_group() to a cpp file.

> 2. Please use a commit message format similar to the previous commits.
>    
>    * module or category: blan blah (all lower case)
>    * ex) foros: delete override keyward from publisher()

I changed every commits that I did. I hope it is correct and helpful.

huchijwk reviewed (changes requested):
Almost done, please align coding style. Fortunately clang format file ('.clang-format') is already included in the project.

There are two methods to check the coding style prior to PR. I did not make a contribution guide and CI workflow yet. sorry. :"-(

**Method 1.** if you are vscode user, set `file` for C_Cpp: Clang_format_path in the preference . And re-format documents you modified.
![image](https://user-images.githubusercontent.com/31328739/196052102-83899ef9-fe1e-479e-b724-d9c88ca617ad.png)

**Method 2.** In the foros top directory, 
```find . -name *.hpp ! -name foros.hpp -o -name *.cpp | xargs clang-format -i -style=file```

 (method 1 + auto formatting is higly recommended)

Eventually, you can see below changes.
```diff
diff --git a/foros/include/akit/failover/foros/cluster_node.hpp b/foros/include/akit/failover/foros/cluster_node.hpp
index 2838634..0fbc6ee 100644
--- a/foros/include/akit/failover/foros/cluster_node.hpp
+++ b/foros/include/akit/failover/foros/cluster_node.hpp
@@ -135,7 +135,8 @@ class ClusterNode : public std::enable_shared_from_this<ClusterNode>,
       rclcpp::CallbackGroupType group_type,
       bool automatically_add_to_executor_with_node = true);
 
-  /// Iterate over the callback groups in the node, calling func on each valid one.
+  /// Iterate over the callback groups in the node, calling func on each valid
+  /// one.
   /**
    * From Humble, get_callback_groups() is replaced with this method.
    * https://github.com/ros2/rclcpp/pull/1723
@@ -144,7 +145,8 @@ class ClusterNode : public std::enable_shared_from_this<ClusterNode>,
    */
   CLUSTER_NODE_PUBLIC
   void for_each_callback_group(
-    const rclcpp::node_interfaces::NodeBaseInterface::CallbackGroupFunction & func);
+      const rclcpp::node_interfaces::NodeBaseInterface::CallbackGroupFunction
+          &func);
 
   /// Create a Publisher.
   /**
diff --git a/foros/src/cluster_node.cpp b/foros/src/cluster_node.cpp
index a9dc831..ce4ba8d 100644
--- a/foros/src/cluster_node.cpp
+++ b/foros/src/cluster_node.cpp
@@ -120,7 +120,8 @@ rclcpp::CallbackGroup::SharedPtr ClusterNode::create_callback_group(
 }
 
 void ClusterNode::for_each_callback_group(
-    const rclcpp::node_interfaces::NodeBaseInterface::CallbackGroupFunction & func) {
+    const rclcpp::node_interfaces::NodeBaseInterface::CallbackGroupFunction
+        &func) {
   node_base_->for_each_callback_group(func);
 }

```

Commit:
ci: applied clang format

Signed-off-by: Taehun Lim <thlim@robotis.com>

routiful commented:
Thank you for details! I checked coding style using clang :) 
Furthermore, I ran `colcon_test` to confirm whether every test are passed. As a result all tests are cleanly passed.

huchijwk reviewed (approved):
LGTM
